drivers: ethernet: Fix adin2111 multiple compatibles support

MaureenHelm opened the pull request:
The adin2111 ethernet driver supports both adi,adin2111 and adi,adin1110
devicetree compatibles, however it failed to build when both compatibles
existed in the same devicetree. This may be an unusual configuration for
real systems, but was found when extending
tests/drivers/build_all/ethernet to cover both compatibles.

Signed-off-by: Maureen Helm <maureen.helm@analog.com>

MaureenHelm committed:
drivers: ethernet: Fix adin2111 multiple compatibles support

The adin2111 ethernet driver supports both adi,adin2111 and adi,adin1110
devicetree compatibles, however it failed to build when both compatibles
existed in the same devicetree. This may be an unusual configuration for
real systems, but was found when extending
tests/drivers/build_all/ethernet to cover both compatibles.

Signed-off-by: Maureen Helm <maureen.helm@analog.com>

MaureenHelm committed:
tests: drivers: build_all: Add adin2111 to ethernet driver build test

Ensures that the eth_adin2111 driver builds for both of its supported
devicetree compatibles by adding adi,adin2111 and adi,adin1110
compatible nodes to the ethernet driver build test.

Sets relevant device device driver initialization priorities to be
equal, allowing devicetree dependency ordinals to determine the
initialization sequence.

Signed-off-by: Maureen Helm <maureen.helm@analog.com>

fabiobaltieri commented on tests/drivers/build_all/ethernet/prj.conf:
```diff
@@ -7,3 +7,9 @@ CONFIG_TEST_RANDOM_GENERATOR=y
 
 CONFIG_GPIO=y
+
+CONFIG_SPI_INIT_PRIORITY=50
+CONFIG_ETH_INIT_PRIORITY=60
+CONFIG_ETH_ADIN2111_INIT_PRIORITY=60
+CONFIG_MDIO_INIT_PRIORITY=70
+CONFIG_PHY_INIT_PRIORITY=80
```
yikes... how about setting all of them at the same value and let the ordinal do the sequencing?

GeorgeCGV replied:
> how about setting all of them at the same value and let the ordinal do the sequencing

Checked such approach, seems to work.

It is definitely possible to get rid of `CONFIG_ETH_ADIN2111_INIT_PRIORITY` in favor of `CONFIG_ETH_INIT_PRIORITY`. Can that be done in this PR or do we need a separate one?

Maybe it is worth to consider modifying `CONFIG_MDIO_INIT_PRIORITY` default to `70`? That should work fine for others since #60816 and #61563.

MaureenHelm replied:
> Checked such approach, seems to work.
> 

Thank you

> It is definitely possible to get rid of `CONFIG_ETH_ADIN2111_INIT_PRIORITY` in favor of `CONFIG_ETH_INIT_PRIORITY`. Can that be done in this PR or do we need a separate one?
> 

I can make the change but I don't have hardware to test it. If you can help, that would be great.

> Maybe it is worth to consider modifying `CONFIG_MDIO_INIT_PRIORITY` default to `70`? That should work fine for others since #60816 and #61563. However, I see that `sam_v71b_xult`, `sam_e70`, `sam4e_xpro` fail.

I was reluctant to change default priorities due to my lack of hardware to test. I need to investigate the sam build failures.

GeorgeCGV reviewed (dismissed):
LGTM for ADIN part.
Non blocking, configuration could be simplified based on @fabiobaltieri input.

However, init priority modification seem to break `sam` builds?

MaureenHelm committed:
drivers: ethernet: Decouple generic MII PHY and adin2111 driver Kconfigs

Decouples the generic MII PHY driver from the adin2111 driver by making
it depend on an ethernet-phy compatible devicetree node rather than the
adin2111 driver not being enabled. This makes it possible to add the
adin2111 driver to tests/drivers/build_all/ethernet

Signed-off-by: Maureen Helm <maureen.helm@analog.com>

MaureenHelm committed:
drivers: ethernet: Remove CONFIG_ETH_ADIN2111_INIT_PRIORITY

Simplifies the adin2111 ethernet driver to use the generic driver class
initialization priority instead of a driver-specific priority.

Suggested-by: Georgij Cernysiov <geo.cgv@gmail.com>
Signed-off-by: Maureen Helm <maureen.helm@analog.com>

GeorgeCGV commented on tests/drivers/build_all/ethernet/prj.conf:
```diff
@@ -7,3 +7,8 @@ CONFIG_TEST_RANDOM_GENERATOR=y
 
 CONFIG_GPIO=y
+
+CONFIG_SPI_INIT_PRIORITY=50
+CONFIG_ETH_INIT_PRIORITY=50
+CONFIG_MDIO_INIT_PRIORITY=50
+CONFIG_PHY_INIT_PRIORITY=50
```
Considering that default priorities are:
* `CONFIG_SPI_INIT_PRIORITY=70`
* `CONFIG_ETH_INIT_PRIORITY=70`
* `CONFIG_PHY_INIT_PRIORITY=70`

It should be enough to use `CONFIG_MDIO_INIT_PRIORITY=70` to make it work for ADIN. I will be able to check it tomorrow.

GeorgeCGV commented on tests/drivers/build_all/ethernet/prj.conf:
```diff
@@ -7,3 +7,8 @@ CONFIG_TEST_RANDOM_GENERATOR=y
 
 CONFIG_GPIO=y
+
+CONFIG_SPI_INIT_PRIORITY=50
+CONFIG_ETH_INIT_PRIORITY=50
+CONFIG_MDIO_INIT_PRIORITY=50
+CONFIG_PHY_INIT_PRIORITY=50
```
Builds and works for ADIN with:

```
CONFIG_MDIO_INIT_PRIORITY=70
CONFIG_ETH_INIT_PRIORITY=70
```

GeorgeCGV replied:
@fabiobaltieri is there anyway to get it cleaner? Should the driver's Kconfig set MDIO and ETH priorities so that it doesn't require further configuration in the app?

fabiobaltieri replied:
Yeah it'd be cool to clean that up, now that we have ordinals in the sequence those priorities do more damage than anything else. I'm working on something more radical to propose soon but in the meantime feel free to change it in the drivers.

GeorgeCGV reviewed (dismissed):
Configuration options were tested. Provided minimal config.
